### ISSUE TITLE
Wildcard ServiceEntry/VirtualService provide surprising access

### ISSUE THREAD

**Issue by rvandegrift** [istio/istio.io#2766]
Using wildcard entries in ServiceEntry and VirtualService can have some surprising interactions.  On an otherwise empty cluster (I'm using GKE 1.10.7 w/Istio 1.0.2) both of the configs below permit access to any host in `googleapis.com` - for instance, I can reach `container.googleapis.com`.  It feels like neither configuration should permit this.

Non-wildcard ServiceEntry with wildcard VirtualService:
```yaml
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: pubsub
spec:
  hosts:
    - pubsub.googleapis.com
  location: MESH_EXTERNAL
  resolution: DNS
  ports:
    - number: 443
      name: https
      protocol: HTTPS
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: star
spec:
  hosts:
    - '*.googleapis.com'
  tls:
    - match:
        - port: 443
          sni_hosts:
            - '*.googleapis.com'
      route:
        - destination:
            host: '*.googleapis.com'
            port:
              number: 443
          weight: 100
```

Wildcard ServiceEntry with non-wildcard VirtualService:
```yaml
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: star
spec:
  hosts:
    - '*.googleapis.com'
  location: MESH_EXTERNAL
  resolution: DNS
  ports:
    - number: 443
      name: https
      protocol: HTTPS

---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: pubsub
spec:
  hosts:
    - pubsub.googleapis.com
  tls:
    - match:
        - port: 443
          sni_hosts:
            - pubsub.googleapis.com
      route:
        - destination:
            host: pubsub.googleapis.com
            port:
              number: 443
          weight: 100
```

**Comment by rvandegrift** [comment]
CC @andraxylia

**Comment by frankbu** [comment]
@rvandegrift I can't reproduce this problem. Maybe the code has been fixed, since you tried it? I'm using the latest code in istio master and both of your configurations rightfully do not allow access to either `pubsub.googleapis.com` or `container.googleapis.com`.

Your first example doesn't work because the ServiceEntry host is `pubsub.googleapis.com` while the destination in the VirtualService is `*.googleapis.com` (there is no such entry in the service registry). If you fix the example, like below, it works as expected (you can curl `pubsub.googleapis.com` but not `pubsub.googleapis.com`).
```
kind: ServiceEntry
metadata:
  name: pubsub
spec:
  hosts:
    - pubsub.googleapis.com
  location: MESH_EXTERNAL
  resolution: DNS
  ports:
    - number: 443
      name: https
      protocol: HTTPS
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: star
spec:
  hosts:
    - '*.googleapis.com'
  tls:
    - match:
        - port: 443
          sni_hosts:
            - 'pubsub.googleapis.com'
      route:
        - destination:
            host: 'pubsub.googleapis.com'
            port:
              number: 443
          weight: 100
```

Your second example doesn't work for the opposite reason (the ServiceEntry host is `*.googleapis.com` but the destination in the VirtualService is `pubsub.googleapis.com`, which again doesn't exist). If you fix it like below, then it will also work as expected.
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: star
spec:
  hosts:
    - '*.googleapis.com'
  location: MESH_EXTERNAL
  resolution: DNS
  ports:
    - number: 443
      name: https
      protocol: HTTPS

---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: pubsub
spec:
  hosts:
    - pubsub.googleapis.com
  tls:
    - match:
        - port: 443
          sni_hosts:
            - pubsub.googleapis.com
      route:
        - destination:
            host: '*.googleapis.com'
            port:
              number: 443
          weight: 100
```

That said, according to the doc, resolution DNS is not supposed to work with wildcard service entries, but it does seem to work just fine when I tried. @rshriram: any idea why?

**Comment by rvandegrift** [comment]
Hi @frankbu - I tried to re-run my test cases on git master, but the helm install isn't working.  I opened #2787 on that, and will retest as soon as I can get master running.

**Comment by rshriram** [comment]
@rvandegrift the configuration for service entry is invalid. We had accidentally disabled service entry validation and forgot to re-enable it for release. Thanks to @frankbu for spotting that.. This should be fixed with https://github.com/istio/istio/pull/9455

**Comment by rshriram** [comment]
Also, *.googleapis.com is a special non standard dns server.. Seems they even have a CNAME entry for *.googleapis.com -- which nobody else does.

**Comment by rvandegrift** [comment]
I see.  That's kind of a bummer, the egress config seems pretty verbose and fragile.  But I don't have a concrete complaint that's appropriate for this issue.

FWIW, I don't think wildcard CNAMEs are rare or unusual.  It's often used it as a service routing mechanism - point the wildcard CNAME to load balanced proxies, use the http Host header to choose the appropriate backend.  I don't know if that's how `*.googleapis.com` works, but it's surely not the only case out there - I've personally done it many times.

**Comment by frankbu** [comment]
@rvandegrift you can still create the ServiceEntry for `*.googleapis.com`, but you need to add an endpoint like this:
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: star
spec:
  hosts:
  - '*.googleapis.com'
  endpoints:
  - address: googleapis.com
  location: MESH_EXTERNAL
  resolution: DNS
  ports:
    - number: 443
      name: https
      protocol: HTTPS
```
Then you can select the specific subdomains that you want to expose using VirtualServices, for example to only allow `pubsub.googleapis.com`:
```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: pubsub
spec:
  hosts:
    - pubsub.googleapis.com
  tls:
    - match:
        - port: 443
          sni_hosts:
            - pubsub.googleapis.com
      route:
        - destination:
            host: '*.googleapis.com'
            port:
              number: 443
          weight: 100
```

**Comment by rvandegrift** [comment]
@frankbu I was able to test with this morning's master, and indeed your config works - but only for the first VirtualService that I create.  If I create a second VirtualService for `container.googleapis.com`, that API fails though `pubsub.googleapis.com` works.

If I create the VirtualService for `container.googleapis.com` first, then it works but `pubsub.googleapis.com` fails.

I got both working by creating separate non-wildcard ServiceEntries for each subdomain, and configuring the VirtualServices to point to the corresponding ServiceEntry.

**Comment by frankbu** [comment]
@rvandegrift I noticed that it does work if instead of creating two virtualservices, you put both hosts in one virtualservice:
```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: pubsub
spec:
  hosts:
    - pubsub.googleapis.com
    - container.googleapis.com
  tls:
    - match:
        - port: 443
          sni_hosts:
            - pubsub.googleapis.com
            - container.googleapis.com
      route:
        - destination:
            host: '*.googleapis.com'
            port:
              number: 443
          weight: 100
```
@rshriram Does this what you would expect?

**Comment by rshriram** [comment]
@frankbu I would have expected both the one you have and the one with two virtual services to work.. That said, I have to look into the code again to refresh my memory as to what me & Isaiah put in..

**Comment by frankbu** [comment]
Fixed in release 1.1 (https://github.com/istio/istio/pull/9570).

@rvandegrift, closing this issue because it should now work with one or two virtual services.

**Comment by rahul6941** [comment]
@rvandegrift 
Please share you configuration . I am trying above but its not working
